### PR TITLE
*: add a metrics for memory trace

### PR DESCRIPTION
**Pull request by rleungx**
Signed-off-by: Ryan Leung <rleungx@gmail.com>

### What problem does this PR solve?

Problem Summary: Add metrics for recording memory usage of each component.

### What is changed and how it works?

What's Changed: Introduce MemoryTraceManager for registering components and flushing according to the specified time interval.

After this PR:
Screen Shot 2021-06-02 at 5 25 34 PM

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/server/src/server.rs [hunk 1]
```diff
@@ -1257,3 +1263,33 @@ impl<R: RaftEngine> EngineMetricsManager<R> {
         }
     }
 }
+
+#[derive(Default)]
```
Comment (5kbpers) [inline components/server/src/server.rs:1267]: How about putting this code to a new `memory.rs`

File: components/server/src/server.rs [hunk 2]
```diff
@@ -1257,3 +1263,33 @@ impl<R: RaftEngine> EngineMetricsManager<R> {
...
+}
+
+impl MemoryTraceManager {
+    pub fn flush(&mut self, _duration: Duration) {
+        for provider in &self.providers {
+            let provider_name = provider.readable_name();
+            let ids = provider.get_children_ids();
+            for id in ids {
+                let sub_trace = provider.sub_trace(id);
+                let sub_trace_name = sub_trace.readable_name();
```
Comment (5kbpers) [inline components/server/src/server.rs:1279]: `readable_name` may not good to be a metric name since words are separated by space.

**Review by 5kbpers — COMMENTED** [review]
rest LGTM

File: components/server/src/server.rs (hunk 2 above)
Comment (5kbpers) [reply components/server/src/server.rs:1279]: Better to add a function like `original_name`?

File: components/server/src/server.rs [hunk 3]
```diff
@@ -927,6 +928,8 @@ impl<ER: RaftEngine> TiKVServer<ER> {
         let mut engine_metrics =
             EngineMetricsManager::new(self.engines.as_ref().unwrap().engines.clone());
         let mut io_metrics = IOMetricsManager::new(fetcher);
+        let mut mem_trace_metrics = MemoryTraceManager::default();
+        mem_trace_metrics.register_provider((&*MEMTRACE_ROOT).to_owned());
```
Comment (nolouch) [inline components/server/src/server.rs:932]: Will we have multiple providers?

**Review by nolouch — COMMENTED** [review]
rest LGTM

File: metrics/grafana/tikv_details.json [hunk 4]
```diff
@@ -3643,6 +3643,96 @@
...
+          "pointradius": 2,
+          "points": false,
+          "renderer": "flot",
+          "seriesOverrides": [],
+          "spaceLength": 10,
+          "stack": false,
+          "steppedLine": false,
+          "targets": [
+            {
+              "expr": "tikv_server_mem_trace_sum{tidb_cluster=\"$tidb_cluster\", instance=~\"$instance\", name=~\"raftstore-.*\"}",
```
Comment (nolouch) [inline metrics/grafana/tikv_details.json:3689]: Need limit to raftstore?

File: components/server/src/server.rs (hunk 3 above)
Comment (rleungx) [reply components/server/src/server.rs:932]: I think so. But for now, we only have one provider.

File: metrics/grafana/tikv_details.json (hunk 4 above)
Comment (rleungx) [reply metrics/grafana/tikv_details.json:3689]: Actually, I put the total component memory and each module of the component together. I think it's ok for now.

**Review by nolouch — APPROVED** [review]
lgtm

**Review by 5kbpers — APPROVED** [review]

File: metrics/grafana/tikv_details.json [hunk 5]
```diff
@@ -3643,6 +3643,96 @@
             "align": false,
             "alignLevel": null
           }
+        },
+        {
+          "aliasColors": {},
+          "bars": false,
+          "dashLength": 10,
+          "dashes": false,
+          "datasource": "playground",
```
Comment (sticnarf) [inline metrics/grafana/tikv_details.json:3652]: The datasource should be replaced with `${DS_TEST-CLUSTER}`

**Pull request merged**